src/App.js

```diff
@@ -78,17 +80,45 @@ const App = () => {
 
   const resetGame = () => {
     // Complete in Wave 4
+    setSquares(generateSquares());
+    setCurrentPlayer(player1);
+    setWinningPlayer('...');
   };
 
+  const onClickCallback = (id) => {
+    if (winningPlayer !== '...') {
+      return;
+    }
+    setSquares((previous) => {
+      return previous.map((row) =>
+        row.map((square) => {
+          if (square.id === id && square.value === '') {
+            square.value = currentPlayer;
+            setCurrentPlayer(currentPlayer === player1 ? player2 : player1);
+            return square;
+          }
+          return square;
+        })
+      );
+    });
+  };
+
+  useEffect(() => {
+    const winner = checkForWinner();
+    if (winner !== null) {
+      setWinningPlayer(winner);
+    }
+  }, [currentPlayer]);
```

Comment:
Good call not calling checkForWinner from the click handler.

As written, checkForWinner should not be called during the click handling. It is written to look at the squares piece of state, but if we are properly calculating a new state and setting it, then the updated squares state wouldn't become observable until the next render. This would have appeared to work, since there is some state cross contamination happening in the click handler, but because you're calling this here, it would still work if you fixed the issue with the square data.

Note, that this doesn't need to be called in an effect. We could call checkForWinner directly in the body of the App component function, similar to how we sometimes run code to calculate a CSS class. It would run each time the board redraws (from updating its state). By using an effect, there's actually a 1 render delay between when the user click occurs, and when the winner is detected. Not noticeable here, but something to be aware of.